src/main/scala/lantern/TensorDifferentiation.scala

```diff
@@ -3150,6 +3186,100 @@ trait TensorDslCudnn extends TensorDslCublas {
+      val res = Tensor(mallocArray[Float](x.scalarCount), x.shape: _*)
+      unchecked[Unit](
+        Seq(s"""
+          |{
+          |cudnnTensorDescriptor_t x_desc;
+          |CUDNN_CALL(cudnnCreateTensorDescriptor(&x_desc));
+          |CUDNN_CALL(cudnnSetTensor4dDescriptor(
+          |    x_desc, CUDNN_TENSOR_NCHW, CUDNN_DATA_FLOAT,
+          |    ${x.shape(0)}, ${x.shape(1)}, ${x.shape(2)}, ${x.shape(3)}));
+          |
+          |cudnnActivationDescriptor_t act_desc;
+          |CUDNN_CALL(cudnnCreateActivationDescriptor(&act_desc));
+          |CUDNN_CALL(cudnnSetActivationDescriptor(act_desc,
+          |                                        /*mode=*/ ${activation.toString},
+          |                                        /*reluNanOpt=*/ CUDNN_PROPAGATE_NAN,
+          |                                        /*relu_coef=*/ 0));
+          |""".stripMargin) ++
+        Seq(
+          "CUDNN_CALL(cudnnActivationForward(\n" +
+          "    cudnnHandle, act_desc,\n" +
+          "    ", one, ", x_desc, ", x.data, ", ", zero, ", x_desc, ", res.data, "));\n" +
+          "}"): _*
+      )
+      res
+    }
+
+    def cudnnActivationBackward(input: TensorR, res: TensorR, activation: Activation.Value): Unit = {
+      assert(input.x.rank == 4, "Currently, activation functions only support tensors of rank 4")
+      assert(input.x.shape == res.x.shape,
+        "Currently, input and result shapes must be equal: ${input.x.shape}, ${res.x.shape}")
+      assert(input.d.shape == res.d.shape,
+        s"Currently, input gradient and result gradient shapes must be equal: ${input.d.shape}, ${res.d.shape}")
+      val zero = NewArray[Float](1); zero(0) = 0
+      val one = NewArray[Float](1); one(0) = 1
+      val inputGrad = Tensor(mallocArray[Float](input.d.scalarCount), input.d.shape: _*)
+      unchecked[Unit](
+        Seq(s"""
+          |{
+          |cudnnTensorDescriptor_t x_desc;
+          |CUDNN_CALL(cudnnCreateTensorDescriptor(&x_desc));
+          |CUDNN_CALL(cudnnSetTensor4dDescriptor(
+          |    x_desc, CUDNN_TENSOR_NCHW, CUDNN_DATA_FLOAT,
+          |    ${input.x.shape(0)}, ${input.x.shape(1)}, ${input.x.shape(2)}, ${input.x.shape(3)}));
+          |
+          |cudnnActivationDescriptor_t act_desc;
+          |CUDNN_CALL(cudnnCreateActivationDescriptor(&act_desc));
+          |CUDNN_CALL(cudnnSetActivationDescriptor(act_desc,
+          |                                        /*mode=*/ ${activation.toString},
+          |                                        /*reluNanOpt=*/ CUDNN_PROPAGATE_NAN,
+          |                                        /*relu_coef=*/ 0));
+          |""".stripMargin) ++
+        Seq(
+          "CUDNN_CALL(cudnnActivationBackward(\n" +
+          "    cudnnHandle, act_desc,\n" +
+          "    ", one, ", x_desc, ", res.x.data, ", x_desc, ", res.d.data, ", x_desc, ", input.x.data, ",\n",
+          "    ", zero, ", x_desc, ", inputGrad.data, "));\n" +
```

Comment:
As suggested by @feiwang3311 : rather than using beta = 0 and `input.d += inputGrad`, use beta = 1.
I'll take on this.